third_party/move/move-vm/runtime/src/storage/dependencies_gas_charging.rs

```diff
@@ -0,0 +1,149 @@
+    gas_meter: &mut impl GasMeter,
+    traversal_context: &mut TraversalContext,
+    serialized_script: &[u8],
+) -> VMResult<()> {
+    let compiled_script = code_storage.deserialize_and_cache_script(serialized_script)?;
+    let compiled_script = traversal_context.referenced_scripts.alloc(compiled_script);
+
+    // TODO(Gas): Should we charge dependency gas for the script itself?
+    check_dependencies_and_charge_gas(
+        code_storage,
+        gas_meter,
+        traversal_context,
+        compiled_script.immediate_dependencies_iter(),
+    )?;
+
+    Ok(())
+}
+
+pub fn check_type_tag_dependencies_and_charge_gas(
+    module_storage: &impl ModuleStorage,
+    gas_meter: &mut impl GasMeter,
+    traversal_context: &mut TraversalContext,
+    ty_tags: &[TypeTag],
+) -> VMResult<()> {
+    // Charge gas based on the distinct ordered module ids.
+    let timer = VM_TIMER.timer_with_label("traverse_ty_tags_for_gas_charging");
+    let ordered_ty_tags = ty_tags
+        .iter()
+        .flat_map(|ty_tag| ty_tag.preorder_traversal_iter())
+        .filter_map(TypeTag::struct_tag)
+        .map(|struct_tag| {
+            let module_id = traversal_context
+                .referenced_module_ids
+                .alloc(struct_tag.module_id());
+            (module_id.address(), module_id.name())
+        })
+        .collect::<BTreeSet<_>>();
+    drop(timer);
+
+    check_dependencies_and_charge_gas(
+        module_storage,
+        gas_meter,
+        traversal_context,
+        ordered_ty_tags,
+    )
+}
+
+/// Traverses the whole transitive closure of dependencies, starting from the specified
+/// modules and performs gas metering.
+///
+/// The traversal follows a depth-first order, with the module itself being visited first,
+/// followed by its dependencies, and finally its friends.
+/// DO NOT CHANGE THE ORDER unless you have a good reason, or otherwise this could introduce
+/// a breaking change to the gas semantics.
+///
+/// This will result in the shallow-loading of the modules -- they will be read from the
+/// storage as bytes and then deserialized, but NOT converted into the runtime representation.
+///
+/// It should also be noted that this is implemented in a way that avoids the cloning of
+/// `ModuleId`, a.k.a. heap allocations, as much as possible, which is critical for
+/// performance.
+pub fn check_dependencies_and_charge_gas<'a, I>(
+    module_storage: &dyn ModuleStorage,
+    gas_meter: &mut impl GasMeter,
+    traversal_context: &mut TraversalContext<'a>,
+    ids: I,
+) -> VMResult<()>
+where
+    I: IntoIterator<Item = (&'a AccountAddress, &'a IdentStr)>,
+    I::IntoIter: DoubleEndedIterator,
+{
+    let _timer = VM_TIMER.timer_with_label("check_dependencies_and_charge_gas");
+
+    // Initialize the work list (stack) and the map of visited modules.
+    //
+    // TODO: Determine the reserved capacity based on the max number of dependencies allowed.
+    let mut stack = Vec::with_capacity(512);
```

Comment:
Why this high capacity? Isn't this wasteful if you only have a few deps?

Reply:
Copied the code from before, I think this was meant to cover for max possible dependency size, though in general there are significantly less dependencies  in practice on average and large allocation only adds overhead.